### PR TITLE
closes #332 enhancement: add switch to maya numbers ...

### LINKED ISSUE #332: Mayan numeral system
The Mayan numeral system was used for two purposes: calculating and calendar.
 Due to the solar year, the Mayan calendar uses a variation of the vigesimal system: the second position ends at 17, the third at position ends at 360.  Subsequent positions use all twenty digits and the place values continue as 18×20×20 = 7,200 and 18×20×20×20 = 144,000, etc. See https://en.wikipedia.org/wiki/Maya_numerals

Hence a switch should be incorrporated to differ between normal and calendar numbers.

### PR DESCRIPTION
**Pull request by t-m-z**
to differ between calandar and other calculations

### CODE REVIEW

**Comment by S-Man42** [comment]
To be honest, I would love to see this as separate tool. We already have a Maya selector, so it could be added as Maya Calendar tool. Moreover, that can be linked into the Date/Calender Section as well :) And this spares the switch option... What do you think?

**Comment by t-m-z** [comment]
Hm – tricky question. If the user could be sure, that the Mayan numbers are uses as an calendar I would agree.
If you do not know whether these symbols ar real vigesimal or calendar-vigesimal a switch would be helpful.

 

Von: MLorenz42 ***@***.***> 
Gesendet: Wednesday, 17 March 2021 22:09
An: S-Man42/GCWizard ***@***.***>
Cc: Thomas ***@***.***>; Author ***@***.***>
Betreff: Re: [S-Man42/GCWizard] #332 enhancement: add switch to maya numbers ... (#402)

 

To be honest, I would love to see this as separate tool. We already have a Maya selector, so it could be added as Maya Calendar tool. Moreover, that can be linked into the Date/Calender Section as well :) And this spares the switch option... What do you think?

**Comment by S-Man42** [comment]
I guess, in doubt he sees both tools and tries both... Hm... I am really no friend of too much options. And in that case for me it's a quite different tool: Maya Numbers and Maya Calendar

**Comment by t-m-z** [comment]
I understand your point but I do not agree in this case.

The Mayan numbering system was used in two ways:
- doing business calculations
- doing calendarian calculations

For calendarian calculations the system was slightly changed at the second position to get a maximum of 360 days.

Additionally think from a user perspective:
you typed a bunch of symbols and realize, you used the wrong tool and have to retype all the symbols. 
What do you feel.

Sent from MailDroid

-----Original Message-----

**Comment by t-m-z** [comment]
another option would be a two-column output: left column vigesimal system right column calendarian system

Sent from MailDroid

-----Original Message-----

File: assets/i18n/de.json [hunk 1]
```diff
@@ -1657,6 +1657,8 @@
   "mayanumbers_online_help_url" : "https://blog.gcwizard.net/manual/de/maya/02-wie-werden-maya-zahlen-ent-bzw-verschlusselt/",
   "mayanumbers_single_numbers" : "Einzelne Zahlen",
   "mayanumbers_vigesimal" : "Zahl im Vigesimalsystem (Zahlenbasis 20)",
+  "mayanumbers_mode_calendar" : "Kalender",
+  "mayanumbers_mode_calculate" : "Sonstig",
```
Comment (S-Man42) [inline assets/i18n/de.json:1661]: Typo. Furthermore, I'd rename it to "Numbers" - and change the order. First (default) numbers, right calender as special case.

File: lib/logic/tools/science_and_technology/numeral_bases.dart [hunk 2]
```diff
@@ -274,3 +278,26 @@ String _doubleDecToBase(double num, int base, String alphabet) {
 
   return out;
 }
+
+String convertDecToMayaCalendar(String input) {
```
Comment (S-Man42) [inline lib/logic/tools/science_and_technology/numeral_bases.dart:282]: Please move this into the maya numbers. This is nothing very common, so it shouldn't be added to the very common numeral_bases. For maya calcs there's already a specific class :)

File: lib/logic/tools/science_and_technology/numeral_bases.dart [hunk 3]
```diff
@@ -2,6 +2,10 @@ import 'dart:math';
 
 const _alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
 
+List<int> mayaSystem = [1, 20, 360, 7200, 144000, 2880000, 57600000, 1152000000, 23040000000, 460800000000, 9216000000000, 184320000000000,3686400000000000];
+Map <String, int> mayaSystemPosition = {'0' : 0, '1' : 1, '2' : 2, '3' : 3, '4' : 4, '5' : 5, '6' : 6, '7' : 7, '9' : 9,
```
Comment (S-Man42) [inline lib/logic/tools/science_and_technology/numeral_bases.dart:6]: IMO, this is not necessary. The _alphabet is already what you need. with convertBase(input, 10, 20) you can convert the base 20 system there and back. (have a look at my maka code... it's a one-liner with only using the already existing tools)

File: assets/i18n/de.json (hunk 1 above)
Comment (t-m-z) [reply assets/i18n/de.json:1661]: done

File: lib/logic/tools/science_and_technology/numeral_bases.dart (hunk 3 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/numeral_bases.dart:6]: basicalle right - but the maya calendar system is slightly different: 
it is not 1, 20, 400, 8000 and so on 
but 1, 20, 360, 720, and so on
Hence convertBase does not work.

File: lib/logic/tools/science_and_technology/numeral_bases.dart (hunk 2 above)
Comment (t-m-z) [reply lib/logic/tools/science_and_technology/numeral_bases.dart:282]: done

**Comment by t-m-z** [comment]
Thanks for your hints, I would still stick to the option switch beacuse these are two kinds of interpreting numeral symbols.
It does not belong to a calendar tool to convert between different systems. Maybe this could be an issue for future work
https://en.wikipedia.org/wiki/List_of_calendars

Regarding a single Mayan number tool maybe there should be an enhancement providing data from
https://www.indianerwww.de/indian/maya-schrift-kalender.htm#:~:text=Dazu%20benutzten%20sie%20das%20Zw%C3%B6lfersystem%2C%20mit%20den%20auch,20%20kin.%2020%20Tage%20bilden%20einen%20Monat%20uinal. or other esoteric web sites.

**Comment by t-m-z** [comment]
I will close this PR and reopen it as a different tool to calculate Mayan calendars.
Hence the output will be according to the 
https://en.wikipedia.org/wiki/Maya_calendar
a date in 
- Long count https://en.wikipedia.org/wiki/Mesoamerican_Long_Count_calendar
- Haab https://en.wikipedia.org/wiki/Haab%CA%BC
- Tzolkin https://en.wikipedia.org/wiki/Tzolk%CA%BCin

**Pull request closed without merging**